File: llm.py
"""LLM extraction with Sarvam-105B primary and Gemini backup."""

from __future__ import annotations

import json
import os
import re

from prompts import EXTRACTION_PROMPT, PRIORITY_PROMPT
from timeouts import run_with_timeout

SARVAM_TIMEOUT_SEC = 12
GEMINI_TIMEOUT_SEC = 20
GEMINI_MODELS = ("gemini-2.5-flash", "gemini-2.0-flash")
SARVAM_CHAT_MODEL = "sarvam-105b"


def parse_json_from_text(text: str) -> dict:
    if not text:
        raise ValueError("Empty model response")
    cleaned = text.strip()
    cleaned = re.sub(r"^```(?:json)?\s*", "", cleaned)
    cleaned = re.sub(r"\s*```$", "", cleaned)
    try:
        parsed = json.loads(cleaned)
        if isinstance(parsed, dict):
            return parsed
        raise ValueError("JSON was not an object")
    except json.JSONDecodeError:
        start = cleaned.find("{")
        end = cleaned.rfind("}")
        if start >= 0 and end > start:
            parsed = json.loads(cleaned[start : end + 1])
            if isinstance(parsed, dict):
                return parsed
        raise


class LLMService:
    def __init__(self, sarvam_client, gemini_api_key: str | None = None):
        self.sarvam = sarvam_client
        raw = gemini_api_key if gemini_api_key is not None else os.getenv("GEMINI_API_KEY")
        self.gemini_api_key = (raw or "").strip().strip('"').strip("'")

    def extract(self, original_text: str, english_text: str, language: str) -> tuple[dict, str]:
        prompt = EXTRACTION_PROMPT.format(
            language=language or "Unknown",
            original_text=original_text or english_text,
            call_text=english_text or original_text,
        )
        return self._complete_json(prompt)

    def assess_priority(self, extraction: dict) -> tuple[dict, str]:
        prompt = PRIORITY_PROMPT.format(
            extraction_json=json.dumps(extraction, ensure_ascii=False, indent=2)
        )
        return self._complete_json(prompt)

    def _complete_json(self, prompt: str) -> tuple[dict, str]:
        errors = []
        try:
            text = self._sarvam_complete(prompt)
            return parse_json_from_text(text), "sarvam-105b"
        except Exception as exc:
            errors.append(f"sarvam: {exc}")
        try:
            text = self._gemini_complete(prompt)
            return parse_json_from_text(text), "gemini"
        except Exception as exc:
            errors.append(f"gemini: {exc}")
            raise RuntimeError(" | ".join(errors)) from exc

    def _sarvam_complete(self, prompt: str) -> str:
        def _call():
            response = self.sarvam.client.chat.completions(
                model=SARVAM_CHAT_MODEL,
                messages=[{"role": "user", "content": prompt}],
                max_tokens=900,
            )
            return _message_text(response)

        return run_with_timeout(_call, SARVAM_TIMEOUT_SEC, "Sarvam chat")

    def _gemini_complete(self, prompt: str) -> str:
        if not self.gemini_api_key:
            raise RuntimeError("GEMINI_API_KEY is not set")

        def _call():
            from google import genai
            from google.genai import types

            client = genai.Client(
                api_key=self.gemini_api_key,
                http_options=types.HttpOptions(timeout=GEMINI_TIMEOUT_SEC * 1000),
            )
            last_error = None
            for model in GEMINI_MODELS:
                try:
                    response = client.models.generate_content(
                        model=model,
                        contents=prompt,
                        config=types.GenerateContentConfig(
                            response_mime_type="application/json",
                            temperature=0.2,
                            automatic_function_calling=types.AutomaticFunctionCallingConfig(
                                disable=True
                            ),
                        ),
                    )
                    text = getattr(response, "text", None)
                    if text:
                        return text
                    last_error = RuntimeError(f"{model} returned empty text")
                except Exception as exc:
                    last_error = exc
            raise RuntimeError(str(last_error) if last_error else "Gemini failed")

        return run_with_timeout(_call, GEMINI_TIMEOUT_SEC, "Gemini")


def _message_text(response) -> str:
    if response is None:
        raise RuntimeError("Empty Sarvam chat response")
    choices = getattr(response, "choices", None)
    if choices:
        message = getattr(choices[0], "message", None)
        if message is not None:
            content = getattr(message, "content", None)
            if isinstance(content, str) and content.strip():
                return content
            if isinstance(content, list):
                parts = []
                for part in content:
                    if isinstance(part, str):
                        parts.append(part)
                    else:
                        parts.append(getattr(part, "text", "") or "")
                joined = "".join(parts).strip()
                if joined:
                    return joined
    if isinstance(response, dict):
        try:
            return response["choices"][0]["message"]["content"]
        except (KeyError, IndexError, TypeError):
            pass
    raise RuntimeError("Could not read Sarvam chat content")

File: services/ai-engine/src/llm.ts
import { parseJsonFromText } from "./json.js";
import { fillExtractionPrompt, fillPriorityPrompt } from "./prompts.js";
import { withTimeout } from "./timeout.js";

const SARVAM_TIMEOUT = 60_000;
const GEMINI_TIMEOUT = 20_000;
const GEMINI_MODELS = ["gemini-2.5-flash", "gemini-2.5-flash-lite"];

function sarvamKey(): string {
  return (process.env.SARVAM_API_KEY ?? "").trim().replace(/^["']|["']$/g, "");
}

function geminiKey(): string {
  return (process.env.GEMINI_API_KEY ?? "").trim().replace(/^["']|["']$/g, "");
}

async function sarvamComplete(prompt: string): Promise<string> {
  const k = sarvamKey();
  if (!k) throw new Error("SARVAM_API_KEY is not set");
  return withTimeout(async () => {
    const res = await fetch("https://api.sarvam.ai/v1/chat/completions", {
      method: "POST",
      // Docs require BOTH headers; subscription key alone hangs. Reasoning is
      // disabled (null) but the model sometimes reasons anyway and truncates
      // at max_tokens with content:null — so the budget covers reasoning +
      // answer (traced live: ~900 reasoning + ~500 answer).
      headers: {
        "Content-Type": "application/json",
        "api-subscription-key": k,
        Authorization: `Bearer ${k}`,
      },
      body: JSON.stringify({
        model: "sarvam-105b",
        messages: [{ role: "user", content: prompt }],
        max_tokens: 2500,
        temperature: 0.2,
        reasoning_effort: null,
        response_format: { type: "json_object" },
      }),
    });
    if (!res.ok) throw new Error(`Sarvam chat failed: ${res.status}`);
    const data: any = await res.json();
    const content = data?.choices?.[0]?.message?.content;
    const text = Array.isArray(content) ? content.map((p: any) => (typeof p === "string" ? p : p?.text ?? "")).join("") : content;
    if (typeof text === "string" && text.trim()) return text;
    throw new Error("Could not read Sarvam chat content");
  }, SARVAM_TIMEOUT, "Sarvam chat");
}

async function geminiComplete(prompt: string): Promise<string> {
  const k = geminiKey();
  if (!k) throw new Error("GEMINI_API_KEY is not set");
  return withTimeout(async () => {
    let lastError: unknown = null;
    for (const model of GEMINI_MODELS) {
      try {
        const res = await fetch(`https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${encodeURIComponent(k)}`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: "application/json", temperature: 0.2 },
          }),
          signal: AbortSignal.timeout(8000),
        });
        if (!res.ok) throw new Error(`${model} failed: ${res.status}`);
        const data: any = await res.json();
        const text = data?.candidates?.[0]?.content?.parts?.map((p: any) => p?.text ?? "").join("");
        if (text) return text;
        lastError = new Error(`${model} returned empty text`);
      } catch (err) {
        lastError = err;
      }
    }
    throw new Error(String(lastError ?? "Gemini failed"));
  }, GEMINI_TIMEOUT, "Gemini");
}

async function completeJson(prompt: string): Promise<{ data: Record<string, unknown>; model: string }> {
  const errors: string[] = [];
  try {
    return { data: parseJsonFromText(await sarvamComplete(prompt)), model: "sarvam-105b" };
  } catch (err) {
    errors.push(`sarvam: ${err}`);
  }
  try {
    return { data: parseJsonFromText(await geminiComplete(prompt)), model: "gemini" };
  } catch (err) {
    errors.push(`gemini: ${err}`);
    throw new Error(errors.join(" | "));
  }
}

export async function extract(originalText: string, englishText: string, language: string) {
  const prompt = fillExtractionPrompt(language || "Unknown", originalText || englishText, englishText || originalText);
  return completeJson(prompt);
}

export async function assessPriority(extraction: unknown) {
  const prompt = fillPriorityPrompt(JSON.stringify(extraction, null, 2));
  return completeJson(prompt);
}
